text/0012-jsonfile.md

```diff
@@ -0,0 +1,60 @@
+# A brief analysis of JSON file-backed storage
+
+Several components in Firefox — including libpref, XULStore, autofill, and logins — follow a general pattern:
+
+- Store data in memory, usually as a JS collection.
+- Load that data from disk during initialization.
+- Persist that data to disk in its entirety, usually serialized as JSON, at some point after each write, and/or at shutdown.
+
+[JSONFile.jsm](https://dxr.mozilla.org/mozilla-central/source/toolkit/modules/JSONFile.jsm) is a centralized implementation of this pattern. It supports Promise-based asynchronous loading, synchronous loading, and flushes asynchronously, defaulting to 1500ms after each change. libpref flushes after 500msec.
+
+This approach is well suited to data that:
+
+- Changes infrequently.
+- Is relatively small.
+- Is only ever directly read or written by a single thread.
+- Needs to be kept entirely in memory for reads.
+- Has a straightforward JSON encoding (encoding binary blobs, such as images, in base64 is relatively inefficient and typically avoided).
+- Has limited durability requirements.
+
+Advantages of this approach:
+- It's relatively simple, with serialization and deserialization being the only steps.
+- Because the file is rewritten frequently, there is some hard-to-measure resilience to gradual file corruption.
+- The file on disk is human-readable and editable, which is convenient for development, test, and support.
+- The in-memory data can be queried and manipulated synchronously.
+- Development is 'natural' for front-end developers — get things working with only in-memory data, then add persistence on top.
+
+Disadvantages:
+- Versioning is often an afterthought: engineers rarely think to version in-memory data formats.
+- Change tracking is almost always an afterthought, thanks to the 'natural' development process: adding syncing later becomes difficult.
+- Users can edit the file on disk, which can result in persistent state that was never created by the component's own logic.
```

Comment:
This is true of any file-based storage format though. I guess you want to say that user are more easily able to edit a text-based format over a binary format.

Reply:
Yep. I'll tweak the wording.